daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/Evaluatable.scala

```diff
@@ -285,7 +285,17 @@ abstract class Evaluatable[+T <: AnyRef](protected val ci: DPathCompileInfo, qNa
   /**
    * Preferred for use in the runtime.
    */
-  @inline final def maybeConstant = constValue_.asInstanceOf[Maybe[T]]
+  @inline final def maybeConstant = {
```

Comment:
In theory, this should just be maybeConstant = constValue_. All we are trying to do here is a cast from T <: AnyRef : Maybe[AnyRef] -> Maybe[T], which is a complete no-op at runtime.

This feels like we should be using DataValuePrimitiveNullable instead of either Maybe type. However, I believe I decided that doing so would involve refactoring all of out expression classes that used T <: AnyRef, which we probably don't want to do.

We can however use the same trick in the Maybe class that I used in the DataValue class to do this conversion with 0 runtime overhead.

Essentially, as part of the Maybe class, define an inline function along the lines of: unsafeCast[T]:Maybe[T] = (Maybe(this.get))

This worked in the case of DataValues because T was purely a phantom type. In the case of Maybe, I think it should be possible to do something similar because of type erasure.

Reply:
Are you sure? ``maybeConst_`` is defined as a ``Maybe[AnyRef]``. Although we know it will always be a ``Maybe[T]``, it's a var and could theoretcially be changed to something other than a ``Maybe[T]``. Seems like the ``unsafeFromMaybeAnyRef`` is more like what we need, which is somewhat similar to the new code.

Re: DataValuePrimitiveNullable, Evaluatables can result in things that are not DataValues, so I'm not sure we could do that.

Reply:
The insight is that T and AnyRef are the same thing, so it should be possible to cast from Maybe[AnyRef] to Maybe[T] without overhead.

In DataValue.unsafeFromMaybeAnyRef, there is added complexity since the actual function is:

```
null -> DataValue.empty
value -> value
```

In the case of Maybe, we have:

```
null -> null
value -> value
```

which can be treated as a single case.

The analogous situation in DataValues, would be getNullablePrimitive:

``` @inline def getNullablePrimitive:DataValue.DataValuePrimitiveNullable = new DataValue(v) ```

which does the cast without doing anything.

Having said that, we are only talking about 1 conditional statement, so it probably isn't worth spending too much time on this.

Reply:
I've just pushed a change that I *think* is what you describe? Does that look right?

Reply:
That looks correct.